fix(discover) Fix negated issue conditions

markstory opened the pull request:
The issue condition should support negation like other conditions. I've changed the semantics a bit on how the issue filter works as it no longer goes into the `group_ids` filter key, but from what I saw snuba can still promote the new condition into a prewhere.

markstory committed:
fix(discover) Fix negated issue conditions

The issue condition should support negation like other conditions. I've
changed the semantics a bit on how the issue filter works as it no
longer goes into the `group_ids` filter key, but from what I saw snuba
can still promote the new condition into a prewhere.

dashed commented on tests/snuba/api/endpoints/test_organization_events_v2.py:
```diff
@@ -1623,6 +1623,49 @@ def test_issue_in_search_and_columns(self):
+            assert response.status_code == 200, response.content
+            data = response.data["data"]
+            assert len(data) == 1
+            assert data[0]["title"] == event1.title
```
@markstory this should be the title of the 2nd event.

markstory replied:
Whoops.

markstory replied:
This test was passing by luck as the two events have the same title :blush: I'll get that fixed.

evanh commented on src/sentry/api/event_search.py:
```diff
@@ -760,6 +760,14 @@ def convert_search_filter_to_snuba_query(search_filter, key=None):
+    elif name == "issue.id":
+        # Handle "has" queries
+        if search_filter.value.raw_value == "":
+            return [["isNull", [name]], search_filter.operator, 1]
```
If group_id is not nullable, why do we do this for has? Can we just hardcode '1' or something to take some load off clickhouse?

markstory replied:
We could hardcode the group_id condition to 0 and skip the isNull call.

evanh commented on tests/sentry/api/test_event_search.py:
```diff
@@ -1185,12 +1185,30 @@ def test_issue_id_filter(self):
+        _filter = get_filter(
+            "!issue:{}".format(group.qualified_short_id), {"organization_id": self.organization.id}
+        )
+        assert _filter.conditions == [["issue.id", "!=", 2]]
```
```suggestion
        assert _filter.conditions == [["issue.id", "!=", group.id]]
```

markstory committed:
Apply suggestions from code review

Co-authored-by: evanh <evanh@users.noreply.github.com>

markstory committed:
Fix has:issue and correct test.

dashed reviewed (approved):
Looks good to me 👍

markstory committed:
Fix !has:issue filter

We still need isNull to prevent the prewhere promotion which excludes
all the results we're interested in.
